Add auto indent functionality

henryprosser opened the pull request:
Option which allows new blocks to be automatically indented to the same level as the previous block

![ScreenRecording2024-08-21at16 50 54-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/aaeb7318-da19-45ad-a32e-7678b2e05dd9)

henryprosser committed:
Add auto indent functionality

henryprosser commented:
Ignore review for now - making changes to the code

sebmeister2077 commented:
Nice, will look more detailed into the PR and test the functionality when you give me the heads up :D

sebmeister2077 commented on src/index.ts:
```diff
@@ -102,14 +107,70 @@ export default class IndentTune implements BlockTune {
+        for (let i = 0; i < mutation.addedNodes.length; i++) {
+            const node = mutation.addedNodes[i]
+            if (node instanceof Element && node.classList.contains('ce-block') && node.getAttribute('data-id') === this.block?.id) {
+                observer.disconnect()
```
Couldn't all this observer logic be simplified to adding 'this.indentCurrentBlockAutomaticallyIfNeeded()' in the constructor and compare previous sibling to know if yes or no?

sebmeister2077 replied:
I think that would reduce complexity and make the code more understandable

sebmeister2077 commented:
Another thing, you should take into consideration that some users wouldn't want their all their block types to be auto indented after the previous indented block, you should change the input interface to maybe something like 
```ts
/**
* enable auto indent if not null or false
*/
autoIndent?: boolean | {
  /**
  * Tunes you want to apply auto indent for.
  * Defaults to all.
  */
  tuneNames?: string[]
}
``` 

  Example scenario for this would be a table block or any media type

henryprosser committed:
Improve auto indent

henryprosser commented:
Thanks for your suggestions. I have made the following changes with the latest commit:

1. Removed mutationObserver code and replaced with calling autoIndentBlock function inside constructor using queueMicrotask. This code is much simpler and hopefully more understandable. I have also moved the function autoIndentBlock() lower down near the indentBlock() and unindentBlock() functions as it makes more sense having it in this area.

2. Based on your recommendation, I have updated the input interface to include a boolean option that enables or disables the autoIndent functionality. I have also added a tuneName array, allowing users to specify only the block types they want to be auto-indented (if the autoIndent has been enabled).

Code is now ready for review :)

henryprosser committed:
Update README with autoIndent parameters

sebmeister2077 commented:
I've tested the code and i've found no issues with the functionality, great job

sebmeister2077 commented:
Hey @henryprosser, i'll wait for your changes regarding the reviews i made

henryprosser commented:
Hi @sebmeister2077, I have already implemented the reviews/changes you suggested with this commit:

> Thanks for your suggestions. I have made the following changes with the latest commit:
> 
> 1. Removed mutationObserver code and replaced with calling autoIndentBlock function inside constructor using queueMicrotask. This code is much simpler and hopefully more understandable. I have also moved the function autoIndentBlock() lower down near the indentBlock() and unindentBlock() functions as it makes more sense having it in this area.
> 2. Based on your recommendation, I have updated the input interface to include a boolean option that enables or disables the autoIndent functionality. I have also added a tuneName array, allowing users to specify only the block types they want to be auto-indented (if the autoIndent has been enabled).
> 
> Code is now ready for review :)

So the code is finished from my perspective, unless there is something specific you are waiting for?

sebmeister2077 commented:
@henryprosser  Yea, i've left some easy to fix comments (3 in total) after you've pushed your latest commit. Do these not appear for you? Let me know if you dont have time for them and i will complete this PR and fix them later on :D
![image](https://github.com/user-attachments/assets/c6c618bc-f38d-433c-955f-43d3d0fd1ac8)

henryprosser commented:
@sebmeister2077 They don't seem to be appearing for me or I can't find them. If you have time to fix them and are happy to then that would be great :)

sebmeister2077 commented:
hey @henryprosser  i got a question, i saw that there is a portion of the code that executes the autoindent when the data is loadded, which woul literally align all blocks to the same indent, would that be expected? i am asking you because you are using this feature in production :D

henryprosser commented:
hi @sebmeister2077 could you point me to this portion of code you are referring to?

sebmeister2077 commented:
> hi @sebmeister2077 could you point me to this portion of code you are referring to?

Yes it is line 119
![image](https://github.com/user-attachments/assets/dc422a06-3c4d-438e-8bcd-0ff9f90cbf0a)
 even tho in the production code it may be on another line.
I am currently creating unit tests for all features and came across this and saw that if i give as input blocks with different indents they all align themselves

henryprosser commented:
Ah, good catch — this behaviour isn’t intended. Ideally, any existing indent levels from loaded data should be preserved. Are you ok to make the necessary changes?

sebmeister2077 commented:
Yes i can make them @henryprosser .
Will prioritise this in the upcoming days and create a new release, fortunately this fix would mean only removal of code :D

sebmeister2077 commented:
Actually i think i will need to dig a bit deeper for a solution to this
